"Daemon" mode

dlespiau opened the pull request:
It'd be nice if we could use hyperstart as a regular daemon instead of just an init system. This allows us to:
- Customise the OS we run inside the VM. There are configurations where we want a fuller OS than just an initrd+hyperstart. In those cases, hyperstart becomes a daemon started by systemd.
- Something else that could be nice is to have hyperstart started by nsenter or systemd-nspawn so we can debug it on the host side (gdb, strace, ...). I have something like this sort of working and this series is the start of it.

laijs commented on src/init.c:
```diff
@@ -1290,7 +1306,8 @@ int main(int argc, char *argv[])
+			 NULL) == -1) {
 		perror("mount devpts failed");
 		return -1;
 	}
```
I think we could move the above code and some init-related code into a function.
and it should be called only when argv[0] == "init".

dlespiau replied:
Oh, sure, I'll have a look.

Currently it's only the system mounts that are causing problems because they are actually trying to remount the already mounted file systems, so might just pull those in a separate function.

laijs replied:
the commit for `Regrouped the early setup code (system mounts, becoming session leader and setting up a controlling terminal in a single function that is skipped when not running as "init".` is missing

laijs commented on src/Makefile.am:
```diff
@@ -1,3 +1,12 @@
 AM_CFLAGS = -Wall -Werror
-bin_PROGRAMS=init
-init_SOURCES=init.c jsmn.c net.c util.c parse.c parson.c container.c exec.c event.c portmapping.c
```
we can compile it to src/hyperstart only.
and copy src/hyperstart to rootfs/init in the make-initrd.sh

dlespiau replied:
Ooops, pushed the wrong branch, the new update should have the correct commmits.

laijs commented:
trivis error:


checking for SYSTEMD... no
configure: error: Package requirements (systemd) were not met:
No package 'systemd' found
Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.
Alternatively, you may set the environment variables SYSTEMD_CFLAGS
and SYSTEMD_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.

dlespiau commented:
PR updated:

- Changed the binary name to "hyperstart" instead of "init" and change its name
  when being copied to the initrd image.
- hyperstart now knows if it has been started as an init system or a regular
  daemon based on argv[0]
- Added a trivial patch to ignore config.h.in~ I had lying around
- Fixed the case where we don't have the systemd .pc file
- Made sure we don't introduce a build-time dependency on pkg-config

dlespiau commented:
PR updated:

- Remove read_cmdline() that wasn't doing anything.
- Regrouped the early setup code (system mounts, becoming session leader and
  setting up a controlling terminal in a single function that is skipped when
  not running as "init".

laijs commented on src/util.c:
```diff
@@ -24,11 +24,6 @@
 #include "container.h"
 #include "../config.h"
 
-char *read_cmdline(void)
```
I'm not sure we should remove it, we may pass the debug level from it in future.
#170

dlespiau replied:
What the function does is "return NULL", I'm not really removing anything useful that would need to be rewritten :)

dlespiau commented:
PR updated:

  - dropped the commit removing read_cmdline()

dlespiau commented:
Is there anything else blocking this PR? Thanks!

dlespiau commented:
ping! :)

devimc commented:
:+1:

dlespiau commented:
Hi,

Can I do anything to help this PR getting merged? Or maybe you do not want to support hyperstart as a regular daemon?

Thanks,

laijs commented on src/init.c:
```diff
@@ -1264,10 +1268,12 @@ static int hyper_loop(void)
 {
-	char *cmdline, *ctl_serial, *tty_serial;
+	if (!is_init)
+		return 0;
```
cloud you move the test up (outside this hyper_setup_init_process())?

and it seams better to remove this global is_init since there is no other usage.

laijs commented on src/init.c:
```diff
@@ -1295,12 +1301,30 @@ int main(int argc, char *argv[])
+		return -1;
+	}
+
+	cmdline = read_cmdline();
```
leave this in the hyper_setup_init_process()

gnawux commented:
@dlespiau Sorry for the delay, we are willing to enable the daemon mode, and we are working on get this to be merged soon. BTW, could you share some tests/comparison of the daemon mode?

Jimmy-Xu commented:
Can one of the admins verify this patch?

Jimmy-Xu commented:
Can one of the admins verify this patch?

laijs commented:
hello @dlespiau, could you update the pr please?

laijs commented:
hyperstart uses devtmpfs for containers' /dev, but devtmpfs is the singleton in the whole system,
it is shared with all the containers and the rest of the system. it is Ok when hyperstart is the init process. is it Ok when hyperstart is daemon mode? if not we should convert it to tmpfs as the same as libcontainer.

dlespiau commented:
Updated PR:

- Rebased onto master (needed to fix a couple of conflicts)
- moved is_init inside main()

dlespiau commented:
@laijs we haven't seen any problems with the containers' /dev so far.
